flink-core/src/main/java/org/apache/flink/api/connector/source/SplitEnumerator.java

```diff
@@ -87,4 +91,18 @@
 	 */
 	@Override
 	default void notifyCheckpointComplete(long checkpointId) throws Exception {}
+
+	/**
+	 * Handles a custom source event from the source reader.
+	 *
+	 * <p>This method has a default implementation that does nothing, because it is only
+	 * required to be implemented by some sources, which have a custom event protocol between
+	 * reader and enumerator. The common events for reader registration and split requests
+	 * are not dispatched to this method, but rather invoke the {@link #addReader(int)} and
+	 * {@link #handleSplitRequest(int, String)} methods.
+	 *
+	 * @param subtaskId the subtask id of the source reader who sent the source event.
+	 * @param sourceEvent the source event from the source reader.
+	 */
+	default void handleSourceEvent(int subtaskId, SourceEvent sourceEvent) {}
```

Comment:
looks like the enumerators will silently ignore these events if this method was forgotten to be overridden? Wondering if a better default would be to either throw unhandled event exception or at the minimum log them so that the user is aware of the issue? The other thing to keep in mind is that the enumerator will only get custom events if the reader has decided to send them which implies that the user is interested in these events in the first place.